runtime/include/sandbox_functions.h

```diff
@@ -28,6 +28,19 @@ sandbox_close_http(struct sandbox *sandbox)
 	client_socket_close(sandbox->client_socket_descriptor, &sandbox->client_address);
 }
 
+/**
+ * Remove the client fd from epoll
+ */
+static inline void
+sandbox_remove_from_epoll(struct sandbox *sandbox)
+{
+        assert(sandbox != NULL);
+
+        int rc = epoll_ctl(worker_thread_epoll_file_descriptor, EPOLL_CTL_DEL, sandbox->client_socket_descriptor, NULL);
```

Comment:
You are right. The code does the same as you said. If the module is in the middle of the chain, it shouldn't be monitored by the epoll, so simply just not adding it to the epoll set. If the module is the first or last module, it will be added to the epoll. For the first module, the client fd should be removed from the epoll when it is finished, but won't be closed. The original sledge will remove the fd from the epoll and close it. For the chain function, we won't close it in the first module but will do in the last module, so we need to differentiate a module is a first module, or a middle module or the last module.

Reply:
My concern here is that it seems that this is being done as part of the inter-sandbox communication, right? We want to do any sandbox/chain setup at the time we add the modules.

Reply:
> My concern here is that it seems that this is being done as part of the inter-sandbox communication, right? We want to do any sandbox/chain setup at the time we add the modules.

Yes, you are right. When a module was added, it will create the server side socket fd and add it to epoll. The fd here is the client fd - the return value of the `accept()`. Sledge also add this fd to epoll to monitor its IN/OUT/ERR event because sledge will read/write data from/to this fd, which can be blocked, but this read/write operation is done within a sandbox. The client fd cannot be created when a module was added.